Mathlib/NumberTheory/ModularForms/JacobiTheta/TwoVariable.lean

```diff
@@ -14,119 +15,501 @@ This file defines the two-variable Jacobi theta function
+    exact (summable_pow_mul_jacobiTheta₂_term_bound S hT 2).mul_left _
+  have hf_sum : Summable fun n : ℤ ↦ f n (z, τ) := by
+    refine (summable_pow_mul_jacobiTheta₂_term_bound S hT 0).of_norm_bounded _ ?_
+    simpa only [pow_zero, one_mul] using norm_jacobiTheta₂_term_le hT hz.le hτ'.le
+  convert hasFDerivAt_tsum_of_isPreconnected hu_sum hVo hVp hf hu hVmem hf_sum hVmem
+
+lemma continuousAt_jacobiTheta₂ (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
+    ContinuousAt (fun p : ℂ × ℂ ↦ jacobiTheta₂ p.1 p.2) (z, τ) :=
+  (hasFDerivAt_jacobiTheta₂ z hτ).continuousAt
+
+/-- Differentiability of `Θ z τ` in `z`, for fixed `τ`. -/
+lemma differentiableAt_jacobiTheta₂_fst (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
+    DifferentiableAt ℂ (jacobiTheta₂ · τ) z :=
+ ((hasFDerivAt_jacobiTheta₂ z hτ).comp z (hasFDerivAt_prod_mk_left z τ)).differentiableAt
+
+/-- Differentiability of `Θ z τ` in `τ`, for fixed `z`. -/
+lemma differentiableAt_jacobiTheta₂_snd (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
+    DifferentiableAt ℂ (jacobiTheta₂ z) τ :=
+  ((hasFDerivAt_jacobiTheta₂ z hτ).comp τ (hasFDerivAt_prod_mk_right z τ)).differentiableAt
+
+lemma hasDerivAt_jacobiTheta₂_fst (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
+    HasDerivAt (jacobiTheta₂ · τ) (jacobiTheta₂' z τ) z := by
+  -- This proof is annoyingly fiddly, because of the need to commute "evaluation at a point"
+  -- through infinite sums of continuous linear maps.
+  have step1 : HasSum (fun n ↦ jacobiTheta₂_term_fderiv n z τ) (jacobiTheta₂_fderiv z τ) :=
+    hasSum_jacobiTheta₂_term_fderiv z hτ
+  let eval_fst_CLM : (ℂ × ℂ →L[ℂ] ℂ) →L[ℂ] ℂ :=
+  { toFun := fun f ↦ f (1, 0), map_add' := by simp, map_smul' := by simp }
+  have step2: HasSum (fun n ↦ (jacobiTheta₂_term_fderiv n z τ) (1, 0))
+      ((jacobiTheta₂_fderiv z τ) (1, 0)) := by convert eval_fst_CLM.hasSum step1
+  have step3 (n : ℤ) : (jacobiTheta₂_term_fderiv n z τ) (1, 0) = jacobiTheta₂'_term n z τ := by
+    simp [jacobiTheta₂'_term, jacobiTheta₂_term_fderiv, jacobiTheta₂_term, mul_comm _ (cexp _)]
+  rw [funext step3] at step2
+  have step4 : HasDerivAt (fun x ↦ jacobiTheta₂ x τ) ((jacobiTheta₂_fderiv z τ) (1, 0)) z :=
+    ((hasFDerivAt_jacobiTheta₂ z hτ).comp z (hasFDerivAt_prod_mk_left z τ)).hasDerivAt
+  rw [← step2.tsum_eq] at step4
+  exact step4
+
+lemma continuousAt_jacobiTheta₂' (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
+    ContinuousAt (fun p : ℂ × ℂ ↦ jacobiTheta₂' p.1 p.2) (z, τ) := by
+  obtain ⟨T, hT, hτ'⟩ := exists_between hτ
+  obtain ⟨S, hz⟩ := exists_gt |im z|
+  let V := {u | |im u| < S} ×ˢ {v | T < im v}
+  have hVo : IsOpen V := ((_root_.continuous_abs.comp continuous_im).isOpen_preimage _
+    isOpen_Iio).prod (continuous_im.isOpen_preimage _ isOpen_Ioi)
+  refine ContinuousOn.continuousAt ?_ (hVo.mem_nhds ⟨hz, hτ'⟩)
+  let u (n : ℤ) : ℝ := 2 * π * |n| * rexp (-π * (T * n ^ 2 - 2 * S * |n|))
+  have hu : Summable u  := by simpa only [u, mul_assoc, pow_one]
+      using (summable_pow_mul_jacobiTheta₂_term_bound S hT 1).mul_left (2 * π)
```

Comment:
There is some repetition here (compare the proof of `hasFDerivAt_jacobiTheta₂`).
Is there a general principle that can be extracted?

Reply:
I'm not sure, I didn't see a nice way of unifying them (without adding extra bulk to the arguments).